### PR TITLE
Task module

### PR DESCRIPTION
**Pull request by sleidig**
Module to manage tasks similar to notes but as separate entities.

- [x] basic structure for list, subrecords list, details properties
- [x] relatedEntities to link any mix of other entities
- [x] task completion as a special work flow and UX
- [x] repetition intervals can be selected in UI + create follow-up tasks automatically on completion of the previous task
- [x] reuse existing date for index instead of recalculating
- [x] code cleanup TodosRelatedToEntity (refactorings across features in separate PR)
- [x] UX TodoDetails & code cleanup (refactorings across features/popups in separate PR --> #921)
- [x] UI feedback for invalid form (--> maybe via refactoring of EntityFormComponent?)
- [x] "Uncomplete" button (undo)
- [x] open details from todo list as popup
- [x] dashboard widget for current tasks
- [x] sort tasks ascendingly by deadline (--> make default sort configurable for entity-list)
- [x] show all fields (including rec interval etc.) in todos opened from the todo list, (maybe define the columns to display directly on the entity config?)
- [ ] filter by date ranges (--> PR #1383 ) for deadline and startDate
- [ ] ... some secret sauce to make this really exciting and inuitive
- ... and reminders as Push/Email notifications (separate issue/PR)

### CODE REVIEW

File: src/app/features/tasks/model/todo.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import { School } from "../../../child-dev-project/schools/model/school";
+import { RecurringActivity } from "../../../child-dev-project/attendance/model/recurring-activity";
+
+@DatabaseEntity("Todo")
+export class Todo extends Entity {
+  static label = $localize`:label for entity:Task`;
+  static labelPlural = $localize`:label (plural) for entity:Tasks`;
+  static toStringAttributes = ["subject"];
+
+  static create(properties: Object): Todo {
```
Comment (TheSlimvReal) [inline src/app/features/tasks/model/todo.ts:32]:
```suggestion
  static create(properties: Partial<Todo>): Todo {
```

File: src/app/features/tasks/recurring-interval/time-interval.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  /** unit (e.g. days, weeks) to be offset */
+  unit: unitOfTime.Base;
+}
+
+export function generateLabelFromInterval(interval: TimeInterval) {
+  return (
+    " " +
+    $localize`:custom interval select option:every ${interval.value} ` +
+    timeunitLabelMap.get(interval.unit)
+  );
```
Comment (TheSlimvReal) [inline src/app/features/tasks/recurring-interval/time-interval.ts:21]:
```suggestion
  return (
    " " +
    $localize`:custom interval select option| e.g. every 2 weeks:every ${interval.value} ${timeunitLabelMap.get(interval.unit)}`
  );
```
This way a re-ordering can also be achieved

File: src/app/features/tasks/recurring-interval/time-interval.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  ["months", $localize`months`],
+  ["month", $localize`months`],
+  ["m", $localize`months`],
+  ["weeks", $localize`weeks`],
+  ["week", $localize`weeks`],
+  ["w", $localize`weeks`],
+  ["days", $localize`days`],
+  ["day", $localize`days`],
+  ["d", $localize`days`],
+]);
```
Comment (TheSlimvReal) [inline src/app/features/tasks/recurring-interval/time-interval.ts:37]: Duplicate of `availableUnits` in the `CustomIntervalComponent`?

File: src/app/features/tasks/tasks-related-to-entity/tasks-related-to-entity.component.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    const designDoc = {
+      _id: "_design/todo_index",
+      views: {
+        by_entity: {
+          map: `(doc) => {
+            if (!doc._id.startsWith("${Todo.ENTITY_TYPE}")) return;
+            if (!Array.isArray(doc.${this.referenceProperty})) return;
+
+            var d = new Date(doc.deadline || null);
+            var dateString = d.getFullYear() + "-" + String(d.getMonth()+1).padStart(2, "0") + "-" + String(d.getDate()).padStart(2, "0")
```
Comment (TheSlimvReal) [inline src/app/features/tasks/tasks-related-to-entity/tasks-related-to-entity.component.ts:59]: Let's use `date-only` and then the date string already has a good format (e.g. `2022-01-29`)

Comment (TheSlimvReal) [reply src/app/features/tasks/tasks-related-to-entity/tasks-related-to-entity.component.ts:59]: Or do you already have a timebound todo in mind?

File: src/app/features/tasks/tasks-related-to-entity/tasks-related-to-entity.component.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      endkey: [entityId],
+      descending: true,
+    });
+  }
+
+  public getNewEntryFunction(): () => Todo {
+    return () => {
+      const newEntry = new Todo();
+      newEntry.relatedEntities = [this.entity.getId(true)];
+      newEntry.assignedTo = []; // TODO: current user
```
Comment (TheSlimvReal) [inline src/app/features/tasks/tasks-related-to-entity/tasks-related-to-entity.component.ts:84]:
```suggestion
      newEntry.assignedTo = [this.sessionService.getCurrentUser()];
```

File: src/app/features/tasks/recurring-interval/custom-interval/custom-interval.component.html [hunk 6]
```diff
@@ -0,0 +1,16 @@
+
+<mat-dialog-content>
+  <span i18n>repeat every</span>
+  <input matInput type="number" [(ngModel)]="selectedValue" class="field-number" [min]="1">
```
Comment (TheSlimvReal) [inline src/app/features/tasks/recurring-interval/custom-interval/custom-interval.component.html:4]: I dont like this component too much and entering something else then a number breaks the code

**Review by TheSlimvReal — COMMENTED** [review]
Looks fine so far. 

I don't like the custom interval component very much. Instead, the filter selection could be 2 dropdown, 1 for number (1-10), one for unit (days, weeks, months, years). This would also make to code much less complicated.

File: src/app/features/tasks/recurring-interval/custom-interval/custom-interval.component.html (hunk 6 above)
Comment (sleidig) [reply src/app/features/tasks/recurring-interval/custom-interval/custom-interval.component.html:4]: I agree it's not terribly pretty yet ... Having a separate popup can be the basis for future extensions to allow more complex intervals, however.

Google Calendar handles it like this:
image

File: src/app/features/tasks/tasks-related-to-entity/tasks-related-to-entity.component.ts (hunk 4 above)
Comment (sleidig) [reply src/app/features/tasks/tasks-related-to-entity/tasks-related-to-entity.component.ts:59]: oh wow, I never thought about the fact that those dates are already stored as string! This is actually copy&paste from other indices but you are right and this should actually speed up the index calculations, I think.

While reminders at a certain time of day might be interesting, I think I would leave this aside in the deadline for now ...

Comment (sleidig) [reply src/app/features/tasks/tasks-related-to-entity/tasks-related-to-entity.component.ts:59]: What about records with an undefined date here? Currently these default to 1970-01-01  shall we set that as a string then?

File: src/app/features/tasks/recurring-interval/display-recurring-interval/display-recurring-interval.component.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import { DynamicComponent } from "../../../../core/view/dynamic-components/dynamic-component.decorator";
+import { ViewDirective } from "../../../../core/entity-components/entity-utils/view-components/view.directive";
+import { ViewPropertyConfig } from "../../../../core/entity-components/entity-list/EntityListConfig";
+import { generateLabelFromInterval, TimeInterval } from "../time-interval";
+
+@DynamicComponent("DisplayRecurringInterval")
+@Component({
+  selector: "app-display-recurring-interval",
+  template: "{{ label }}",
+  changeDetection: ChangeDetectionStrategy.OnPush,
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/features/tasks/recurring-interval/display-recurring-interval/display-recurring-interval.component.ts:11]: As far as I understand OnPush would only trigger on Input changes, so in this case only at the start. Seems to work correctly, but maybe we should leave the move to OnPush aside now and do some benchmarking + change it for all EditComponents consistently

File: src/app/features/tasks/recurring-interval/time-interval.ts (hunk 3 above)
Comment (sleidig) [reply src/app/features/tasks/recurring-interval/time-interval.ts:36]: :+1:

File: src/app/features/tasks/recurring-interval/display-recurring-interval/display-recurring-interval.component.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/features/tasks/recurring-interval/display-recurring-interval/display-recurring-interval.component.ts:11]: #1586

File: src/app/features/tasks/recurring-interval/custom-interval/custom-interval.component.html (hunk 6 above)
Comment (sleidig) [reply src/app/features/tasks/recurring-interval/custom-interval/custom-interval.component.html:4]: I guess only Chrome prevents text from being entered. Added a validation/transformation now.

File: src/app/features/todos/todo-details/todo-details.component.html [hunk 8]
```diff
@@ -0,0 +1,56 @@
+<app-dialog-close mat-dialog-close></app-dialog-close>
+
+<mat-dialog-content>
+  <h1>{{ entity.subject }}</h1>
+
+  <div class="margin-bottom-large">
+    <app-todo-completion [entity]="entity"></app-todo-completion>
```
Comment (TheSlimvReal) [inline src/app/features/todos/todo-details/todo-details.component.html:7]: This should not be visible when creating a new entity

**Review by TheSlimvReal — COMMENTED** [review]
Works already very well! Nice works.

Some bugs I discovered:

- [x] Changing repeat value and clicking complete without saving in between, shows popup for changes on server
- [x] Using a custom interval still leads to "every undefined weeks/years/months/days"
- [x] The different custom intervals which have been created are persisted throughout a session. Is this intended?
- [x] I think clicking the completed button should also close the popup (maybe after a few seconds)

**Review by TheSlimvReal — CHANGES_REQUESTED** [review]
See comments above

File: src/app/core/demo-data/demo-data.service.ts [hunk 9]
```diff
@@ -91,7 +91,7 @@ export class DemoDataService {
     // save the generated data
     for (const generator of this.dataGenerators) {
       await this.entityMapper.saveAll(generator.entities);
-      // Wait for other async tasks in the queue e.g. ConfigService setting up config after it has been saved
+      // Wait for other async todos in the queue e.g. ConfigService setting up config after it has been saved
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/demo-data/demo-data.service.ts:94]: ups, Webstorm was a bit overly active here (and I didn't properly review it) :+1:

File: src/app/features/todos/recurring-interval/edit-recurring-interval/edit-recurring-interval.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    expect(
+      component.predefinedIntervals.find((o) => o.interval === customInterval)
+    ).toBeTruthy();
+  });
+
+  it("should open dialog for custom interval and select the result", () => {
+    const customIntervalResult: TimeInterval = {amount: 88, unit: "days"};
+    mockDialog.open.and.returnValue({
+      afterClosed: () => of(customIntervalResult),
+    } as MatDialogRef<typeof component>);
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/features/todos/recurring-interval/edit-recurring-interval/edit-recurring-interval.component.spec.ts:54]: thanks

File: src/app/core/form-dialog/form-dialog.service.ts [hunk 11]
```diff
@@ -83,22 +83,24 @@ export class FormDialogService {
    * Open a form in a popup that allows to edit the given entity.
    * @param entity
    * @param columnsOverall
+   * @param component
    */
   openSimpleForm<E extends Entity>(
     entity: E,
-    columnsOverall: ColumnConfig[]
+    columnsOverall: ColumnConfig[],
+    component?: ComponentType<any>
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/form-dialog/form-dialog.service.ts:91]: to be resolved with refactoring of #921

File: src/app/core/entity-components/entity-form/entity-form/entity-form.component.ts [hunk 12]
```diff
@@ -33,6 +33,9 @@ import { ConfirmationDialogService } from "../../../confirmation-dialog/confirma
   encapsulation: ViewEncapsulation.None,
 })
 export class EntityFormComponent<T extends Entity = Entity> implements OnInit {
+  // TODO: move the buttons into FormComponent and make this a plain generated <form> for easier reuse
+  @Input() noButtons = false;
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/entity-components/entity-form/entity-form/entity-form.component.ts:37]: I wanted to make this very explicit that the default is to have buttons and this option is more of a workaround to repurpose the component.
Let's clean this up more properly in #921 by ideally splitting buttons and form into separate components and merging this with RowDetails

File: src/app/core/entity/database-indexing/database-indexing.service.ts [hunk 13]
```diff
@@ -81,6 +81,60 @@ export class DatabaseIndexingService {
...
+   * generateIndexOnProperty("myIndex", "Note", "category", false);
+   * queryIndexDocs(Note, "myIndex/by_category")
+   *
+   * @param indexId id to query this index after creation (--> {indexId}/by_{referenceProperty})
+   * @param entityTypePrefix entity type to limit the documents included in this index
+   * @param referenceProperty property key on the documents whose value is indexed as a query key
+   * @param isArrayProperty whether referenceProperty is an array of values to be indexed separately or a single value
+   * @param secondaryIndex (optional) additional property to emit as a secondary index to narrow queries further
+   */
+  generateIndexOnProperty(
```
Comment (sleidig) [inline src/app/core/entity/database-indexing/database-indexing.service.ts:100]: one step towards indices generated from schema (#260, #262)

**Review by TheSlimvReal — APPROVED** [review]
Great look and feel. Working with it I don't see any bugs or issues. Code wise also looks good, however test coverage could be better 😄 
I made some styling improvements already and also left some small remarks.

File: src/app/features/todos/todo-details/todo-details.component.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        this.alertService.addDanger(err.message);
+      }
+    }
+  }
+
+  async completeTodo() {
+    if (this.form.dirty) {
+      const confirmationResult = await this.confirmationDialog.getConfirmation(
+        $localize`Save changes?`,
+        $localize`Do you want to save your changes to the ${Todo.label} before marking it as completed? Otherwise, changes will be discarded.`,
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/features/todos/todo-details/todo-details.component.ts:79]: that's kinda true ... from a UX rule to do nothing unexpected I am a bit reluctant to make it an inherent save button (it's in a totally different place - and more important: it closes the form) but you are probably right that there is hardly a reasonable scenario where you wouldn't wanna save.

**Pull request merged**